File: modules/student-experience/src/main/java/com/campusos/portal/service/ClassKey.java
package com.campusos.portal.service;

import com.campusos.portal.domain.Student;
import com.campusos.portal.domain.TeachingAssignment;

/**
 * A class: one section of one semester of one department. This is the unit a faculty member
 * is assigned to teach, and the unit a student belongs to — Student already carries all three
 * columns, so the scope rule joins on data that exists.
 */
public record ClassKey(String department, int semester, String section) {

    public ClassKey {
        if (department == null || department.isBlank())
            throw new IllegalStateException("department missing — class scope refused");
        if (section == null || section.isBlank())
            throw new IllegalStateException("section missing — class scope refused");
    }

    public static ClassKey of(Student s) {
        return new ClassKey(s.department, s.semester, s.section);
    }

    public static ClassKey of(TeachingAssignment a) {
        return new ClassKey(a.department, a.semester, a.section);
    }

    public boolean matches(Student s) {
        return department.equals(s.department) && semester == s.semester && section.equals(s.section);
    }

    /** Stable round-trippable form for a form field, so no free-text class ever reaches a query. */
    public String token() {
        return department + "|" + semester + "|" + section;
    }

    public static ClassKey parse(String token) {
        String[] p = token == null ? new String[0] : token.split("\\|");
        if (p.length != 3) throw new IllegalStateException("unreadable class token");
        try {
            return new ClassKey(p[0], Integer.parseInt(p[1]), p[2]);
        } catch (NumberFormatException e) {
            throw new IllegalStateException("unreadable class token");
        }
    }

    public String label() {
        return department + " · Semester " + semester + " · Section " + section;
    }
}

File: modules/student-experience-js/src/service/ClassKey.js
/**
 * A class: one section of one semester of one department. This is the unit a faculty member is
 * assigned to teach, and the unit a student belongs to.
 */
export class ClassKey {
  constructor(department, semester, section) {
    if (!department || !department.trim()) throw new Error('department missing — class scope refused');
    if (!section || !section.trim()) throw new Error('section missing — class scope refused');
    this.department = department;
    this.semester = semester;
    this.section = section;
    Object.freeze(this);
  }

  static of(entity) { return new ClassKey(entity.department, entity.semester, entity.section); }

  matches(s) {
    return this.department === s.department && this.semester === s.semester && this.section === s.section;
  }

  equals(other) {
    return !!other && this.department === other.department
      && this.semester === other.semester && this.section === other.section;
  }

  /** Stable round-trippable form for a form field, so no free-text class ever reaches a query. */
  token() { return `${this.department}|${this.semester}|${this.section}`; }

  static parse(token) {
    const p = (token ?? '').split('|');
    if (p.length !== 3) throw new Error('unreadable class token');
    const semester = Number.parseInt(p[1], 10);
    if (Number.isNaN(semester)) throw new Error('unreadable class token');
    return new ClassKey(p[0], semester, p[2]);
  }

  label() { return `${this.department} · Semester ${this.semester} · Section ${this.section}`; }
}
